src/cmdline/mod.rs

```diff
@@ -336,172 +250,641 @@ impl Cmdline {
+/// # Examples
+///
+/// ```rust
+/// # use linux_loader::cmdline::*;
+/// # use std::ffi::CString;
+/// let mut cl = Cmdline::new(100).unwrap();
+/// cl.boot_args_mut().insert_str("foobar").unwrap();
+/// let cl_cstring = CString::new(cl.as_str().unwrap()).unwrap();
+/// assert_eq!(cl_cstring.to_str().unwrap(), "foobar");
+/// ```
+#[derive(Clone, Debug)]
+pub struct Cmdline {
+    boot_args: BootArgs,
+    init_args: InitArgs,
+    cmdline_capacity: usize,
+}
+
+impl Cmdline {
+    /// Constructs an empty [`Cmdline`] with the given capacity, including the nul terminator and ' -- ' delimiter
+    /// between boot args and init args.
+    ///
+    /// # Arguments
+    ///
+    /// * `cmdline_capacity` - Command line capacity. Must be greater than 0.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let cl = Cmdline::new(100);
+    /// ```
+    /// [`Cmdline`]: struct.Cmdline.html
+    pub fn new(cmdline_capacity: usize) -> Result<Cmdline> {
+        if cmdline_capacity == 0 {
+            return Err(Error::InvalidCapacity);
+        }
+
+        Ok(Cmdline {
+            boot_args: BootArgs(String::new()),
+            init_args: InitArgs(String::new()),
+            cmdline_capacity,
+        })
+    }
+
+    /// Returns the BootArgs representation of the command line.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new(100).unwrap();
+    /// assert!(cl.boot_args_mut().insert_key_value("a", "b").is_ok());
+    /// assert_eq!(cl.as_str().unwrap(), "a=b");
+    /// ```
+    pub fn boot_args_mut(&mut self) -> &mut BootArgs {
+        &mut self.boot_args
+    }
+
+    /// Returns the InitArgs representation of the command line.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new(100).unwrap();
+    /// assert!(cl.boot_args_mut().insert_key_value("a", "b").is_ok());
+    /// assert!(cl.init_args_mut().insert_key_value("c", "d").is_ok());
+    /// assert_eq!(cl.as_str().unwrap(), "a=b -- c=d");
+    /// ```
+    pub fn init_args_mut(&mut self) -> &mut InitArgs {
+        &mut self.init_args
+    }
+
+    /// Returns the string representation of the command line without the nul terminator.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new(100).unwrap();
+    /// assert!(cl.boot_args_mut().insert_str("foo").is_ok());
+    /// assert!(cl.init_args_mut().insert_str("bar").is_ok());
+    /// assert_eq!(cl.as_str().unwrap(), "foo -- bar");
+    /// ```
+    pub fn as_str(&self) -> Result<String> {
+        if self.boot_args.0.is_empty() {
+            return Err(Error::NoBootArgsInserted);
+        }
+
+        if self.init_args.0.is_empty() {
+            if self.boot_args.0.len() + 1 > self.cmdline_capacity {
+                // need space for the null terminator
```

Comment:
In the documentation comment we're saying that we're not adding the nul terminator, which one is the correct comment?

Reply:
We don't add any '\0' char per se when returning the String from as_str() method. However, we need to consider the space required for the nul terminator within the cmdline_capacity so that's why I wrote the if condition adding + 1 (and added a comment to be more clear) - actually it's the same condition used as before, but written in another way.

Reply:
I'll come back to the old condition and remove the comment.

Reply:
But your questions triggered another question in my mind: who should be responsible for adding the '\0' char to the cmdline ? I'm asking because from what I understand, currently, nobody does this... I checked the linux-loader/loader crate and the function load_cmdline() does not do this; it just takes the cmdline.as_str() turns it into bytes and then writes these bytes into guest's memory; is my understanding right? shouldn't we add a '\0' at the end in as_str() method to make sure the cmdline is nul terminated?

Reply:
We should be using CString to get a null terminated string.